### ISSUE TITLE
More probability distributions?

### ISSUE THREAD

**Issue by freddiev4** [pytorch/pytorch#3185]
Looking at the [contributing](https://github.com/pytorch/pytorch/blob/master/CONTRIBUTING.md) guide and I'm currently here 🙂 

![image](https://user-images.githubusercontent.com/8339018/31792467-2e184fbe-b4ea-11e7-8d23-b62adb1a268e.png)

I recently saw this PR here on PyTorch https://github.com/pytorch/pytorch/pull/3165 and was wondering if you're looking for more probability distributions to be added e.g. Poisson, Exponential, Geometric.

If so, I'd like to open up discussion about design/implementation of these distributions and which ones would be useful to add.

**Comment by FreddieV4** [comment]
FWIW, I just saw the link to the other pytorch discussion channels. Should I move this to one of those other channels?

**Comment by apaszke** [comment]
Yes! We're definitely open to adding more distributions. #3165 is a PR that bootstrapped the whole thing, so if you're interested then please send code!

We're going to follow the `tf.distributions` API. To start it's best to either add new distributions or methods like `sample_n`, `mean`, `mode`, `var`, `pdf`, `cdf`, `log_pdf`, `log_cdf`.

Let's keep the discussion here.

**Comment by FreddieV4** [comment]
Awesome. Sounds good to me! Will look into adding some new methods/distributions over the next week or so and open a PR (or two)

**Comment by Kaixhin** [comment]
Thanks for offering to contribute! This has been long awaited (see https://github.com/pytorch/pytorch/issues/678), but by settling on the distributions API there should be more clarity on how to move forward with this. It seems that, following the new PR, having `sample` and `log_prob` for a few distributions (exponential, uniform, categorical, beta) would be the most useful, and the moving onto filling out some more methods.

**Comment by rachtsingh** [comment]
I'm happy to help as well! To clarify - what source should we use for implementations of the special logpdf? Not sure about licensing issues.

I can tackle the Beta distribution - I recently implemented ~80% of this in CUDA already so it's just a matter of cleaning it up for the new API and thinking about tests / stability.

**Comment by alykhantejani** [comment]
@apaszke @colesbury if no one is working on adding `sample_n` right now @geoffroeder can send a PR as he's already started implementing it and was looking at ways to contribute more

**Comment by jwvdm** [comment]
We've been working on a library for deep generative models that extends PyTorch. We have currently written implementations for the following distributions (and have been adding distributions almost daily)

- Normal 
- Concrete (a.k.a. Gumbel-Softmax)
- Exponential
- Laplace
- Logistic
- Uniform

Note that all of these distributions are "reparameterized" distributions for which sampled values are differentiable. 

We wrote our distribution API before @colesbury's changes last week. In particular, we implement a a sample method `Distribution.sample(self, *sizes)` that accepts an optional arguments to draw multiple samples. 

We'd be happy to prepare a pull request.

**Comment by Kaixhin** [comment]
@jwvdm That sounds fantastic - please submit a PR. Whilst we are currently basing the API on TensorFlow, there's no need to follow it exactly if alterations make sense for PyTorch.

Having reparameterised distributions readily available sounds convenient, and indeed [what TensorFlow provides already](https://www.tensorflow.org/api_docs/python/tf/distributions/ReparameterizationType). Having a  section in the docs on a) the difference between reparameterised and normal/non-reparameterised distributions and b) what can be done with reparameterised distributions would be good.

**Comment by jwvdm** [comment]
@Kaixhin Sounds good – we'll get on it! I think that what we'll do is issue pull requests one at a time. We might need a couple of days to put the first one together. We used numpy and scipy to implement our unit tests, so we will need to eliminate those dependencies before we can issue a pull request.

**Comment by fritzo** [comment]
My team has been working on a distributions library on top of Pytorch, and we'd be happy to move that code upstream. We have implemented: `Bernoulli`, `Beta`, `Categorical`, `Cauchy`, `Delta`, `Dirichlet`, `Exponential`, `Gamma`, `HalfCauchy`, `LogNormal`, `Multinomial`, `Normal` (univariate, multivariate, and multivariate w/ Cholesky), `Poisson`, and `Uniform`. Some of these just use `scipy` under the hood, and some only have cpu implementations. We're hoping to open source next week.

**Comment by gokceneraslan** [comment]
@fritzo how did you implement lgamma/digamma/beta functions?

**Comment by fritzo** [comment]
@gokceneraslan we implemented a `log_gamma()` function on top of pytorch. We haven't written any Cuda or C++ code.

**Comment by freddiev4** [comment]
Looks like there's a lot of implementations people have already been working on. You all should open PRs with some of the stuff you've worked on; I can work on contributing in some other way 🙂

**Comment by gokceneraslan** [comment]
lgamma is already in (i.e. torch.lgamma()) but without the derivative, so I had to use this to do backward() on it:

```python
from torch.autograd import Function
import torch

from scipy.special import digamma

class lgamma(Function):
    def forward(self, input):
        self.save_for_backward(input)
        return torch.lgamma(input)

    def backward(self, grad_output):
        input, = self.saved_tensors
        res = torch.DoubleTensor(digamma(input.numpy())).type_as(input)
        return grad_output*res


```

**Comment by fritzo** [comment]
My team just open source a [distributions library](https://github.com/uber/pyro/tree/dev/pyro/distributions) as part of the [Pyro PPL](http://pyro.ai). Here is our [`log_gamma()`](https://github.com/uber/pyro/blob/dev/pyro/distributions/util.py#L8) implementation. We plan to submit PRs soon to move this code upstream into PyTorch.

**Comment by stepelu** [comment]
@fritzo I understand for the `log_gamma()` you're referring to https://github.com/uber/pyro/blob/b94f06a2257bf362208c28048544deb297aa5e24/pyro/distributions/util.py#L8 ?
I have a reasonaly accurate implementation (not sure if that's actually needed for float32) at: https://github.com/stepelu/lua-sci/blob/e43411044e45840230e9d5741e5c6a9960d443ed/math.lua#L128 
Lua code I'm afraid, should be easy to read though, and the cited PhD thesis in the comment is relevant.
I am concerned about the numerical accuracy/stability of this approach though (and that's indeed why I didn't end up implement the Gamma distribution in my small repo), as the numerical guarantees of the function's approximation don't transfer to its derivative.
Have you done some tests in this direction?

**Comment by gokceneraslan** [comment]
@stepelu @fritzo I filed a bug report to Pyro (https://github.com/uber/pyro/issues/567). Let's discuss it there.

**Comment by Kaixhin** [comment]
Think we can close this one now and if people want to contribute specific distributions now they can open up new issues/PRs.
